app/forms/declaration/BorderTransport.scala

```diff
@@ -34,21 +35,118 @@ object BorderTransport extends DeclarationPage {
-      .verifying("supplementary.transportInfo.meansOfTransport.CrossingTheBorder.IDNumber.error.invalid", isAlphanumericWithAllowedSpecialCharacters)
-  )(BorderTransport.apply)(BorderTransport.unapply)
+    "borderTransportType" -> requiredRadio("declaration.transportInformation.meansOfTransport.crossingTheBorder.error.empty")
+      .verifying(
+        "declaration.transportInformation.meansOfTransport.crossingTheBorder.error.incorrect",
+        isContainedIn(allowedMeansOfTransportTypeCodes)
+      ),
+    transportReferenceMapping("IMOShipIDNumber", IMOShipIDNumber),
+    transportReferenceMapping("nameOfVessel", NameOfVessel),
+    transportReferenceMapping("wagonNumber", WagonNumber),
+    transportReferenceMapping("vehicleRegistrationNumber", VehicleRegistrationNumber),
+    transportReferenceMapping("IATAFlightNumber", IATAFlightNumber),
+    transportReferenceMapping("aircraftRegistrationNumber", AircraftRegistrationNumber),
+    transportReferenceMapping("europeanVesselIDNumber", EuropeanVesselIDNumber),
+    transportReferenceMapping("nameOfInlandWaterwayVessel", NameOfInlandWaterwayVessel)
+  )(form2Model)(model2Form)
 
   def form(): Form[BorderTransport] = Form(BorderTransport.formMapping)
+
+  private def form2Model: (
+    Option[String],
+    String,
+    Option[String],
+    Option[String],
+    Option[String],
+    Option[String],
+    Option[String],
+    Option[String],
+    Option[String],
+    Option[String]
+  ) => BorderTransport = {
+
+    case (
+        nationality,
+        transportType,
+        shipIdNumber,
+        nameOfVessel,
+        wagonNumber,
+        vehicleRegistrationNumber,
+        flightNumber,
+        aircraftRegistrationNumber,
+        europeanVesselIDNumber,
+        nameOfInlandWaterwayVessel
+        ) =>
+      BorderTransport(
+        nationality,
+        transportType,
+        form2Ref(
+          shipIdNumber,
+          nameOfVessel,
+          wagonNumber,
+          vehicleRegistrationNumber,
+          flightNumber,
+          aircraftRegistrationNumber,
+          europeanVesselIDNumber,
+          nameOfInlandWaterwayVessel
+        )
+      )
+  }
+
+  private def model2Form: BorderTransport => Option[
+    (
+      Option[String],
+      String,
+      Option[String],
+      Option[String],
+      Option[String],
+      Option[String],
+      Option[String],
+      Option[String],
+      Option[String],
+      Option[String]
+    )
+  ] =
+    implicit model =>
+      Some(
+        model.meansOfTransportCrossingTheBorderNationality,
+        model.meansOfTransportCrossingTheBorderType,
+        model2Ref(IMOShipIDNumber),
+        model2Ref(NameOfVessel),
+        model2Ref(WagonNumber),
+        model2Ref(VehicleRegistrationNumber),
+        model2Ref(IATAFlightNumber),
+        model2Ref(AircraftRegistrationNumber),
+        model2Ref(EuropeanVesselIDNumber),
+        model2Ref(NameOfInlandWaterwayVessel)
+    )
```

Comment:
Why logic here is so complicated?

Reply:
Yes it is a bit fiddly.  It's because we have 8 different input fields now - all optional except for the one that is associated with the selected radio option.

I started by having all 8 bound to the same form field.  That works but when you change radio option the previously typed value is visible.  Plus the validation for each type field might be different (it isn't at the moment but that could change)